File: Languages/SyntaxGenDotNet.CSharp/Utilities/AttributeUtility.cs
using System.Collections.ObjectModel;
using System.Linq.Expressions;
using System.Reflection;
using SyntaxGenDotNet.Attributes;
using SyntaxGenDotNet.Syntax;
using SyntaxGenDotNet.Syntax.Tokens;

namespace SyntaxGenDotNet.CSharp.Utilities;

/// <summary>
///     Provides utility methods for working with attributes in the C# language.
/// </summary>
public static class AttributeUtility
{
    /// <summary>
    ///     Writes a custom attribute to the specified node.
    /// </summary>
    /// <param name="target">The node to which to write the attribute.</param>
    /// <param name="attributeExpression">The attribute expression to write.</param>
    public static void WriteCustomAttribute(SyntaxNode target, MemberInitExpression attributeExpression)
    {
        var options = new TypeWriteOptions {TrimAttributeSuffix = true, WriteNamespace = true, WriteAlias = false};

        // explicit creation of the open bracket token to avoid the trailing whitespace
        // of the previous token being trimmed, as Operators.OpenBracket defaults to trimming.
        var openBracket = new OperatorToken(Operators.OpenBracket.Text, false);

        target.AddChild(openBracket);
        TypeUtility.WriteAlias(target, attributeExpression.Type, options);

        ReadOnlyCollection<Expression> arguments = attributeExpression.NewExpression.Arguments;
        bool hasArguments = arguments.Count > 0;
        bool hasBindings = attributeExpression.Bindings.Count > 0;

        if (hasArguments || hasBindings)
        {
            target.AddChild(Operators.OpenParenthesis);

            if (hasArguments)
            {
                WriteArguments(target, arguments);
            }

            if (hasBindings)
            {
                WriteBindings(hasArguments, target, attributeExpression, options);
            }

            target.AddChild(Operators.CloseParenthesis);
        }

        target.AddChild(Operators.CloseBracket.With(o => o.TrailingWhitespace = WhitespaceTrivia.NewLine));
    }

    /// <summary>
    ///     Writes all known supported custom attribute to the specified node.
    /// </summary>
    /// <param name="generator">The syntax generator.</param>
    /// <param name="target">The node to which to write the attributes.</param>
    /// <param name="member">The member whose custom attributes to write.</param>
    public static void WriteCustomAttributes(SyntaxGenerator generator, SyntaxNode target, MemberInfo member)
    {
        foreach (AttributeExpressionWriter writer in generator.AttributeExpressionWriters)
        {
            Type attributeType = writer.AttributeType;
            Attribute[] attributes = member.GetCustomAttributes(attributeType, false).OfType<Attribute>().ToArray();
            IEnumerable<Expression> expressions = writer.CreateAttributeExpressions(member, attributes);

            foreach (Expression expression in expressions)
            {
                if (expression is MemberInitExpression memberInitExpression)
                {
                    WriteCustomAttribute(target, memberInitExpression);
                }
            }
        }
    }

    private static void WriteArguments(SyntaxNode target, IReadOnlyList<Expression> arguments)
    {
        for (var index = 0; index < arguments.Count; index++)
        {
            Expression argument = arguments[index];
            WriteResolvedExpression(target, argument);
        }
    }

    private static void WriteBindings(bool writeComma,
        SyntaxNode target,
        MemberInitExpression expression,
        TypeWriteOptions options)
    {
        SyntaxNode comma = Operators.Comma.With(o => o.TrailingWhitespace = " ");
        ReadOnlyCollection<MemberBinding> bindings = expression.Bindings;
        options.WriteNamespace = false;

        for (var index = 0; index < bindings.Count; index++)
        {
            MemberBinding binding = bindings[index];
            if (binding is not MemberAssignment assignment)
            {
                continue;
            }

            if (index > 0 || writeComma)
            {
                target.AddChild(comma);
            }

            target.AddChild(new IdentifierToken(binding.Member.Name));
            target.AddChild(Operators.Assignment);
            if (assignment.Expression.Type.IsEnum)
            {
                TypeUtility.WriteAlias(target, assignment.Expression.Type, options);
                target.AddChild(Operators.Dot);
            }

            target.AddChild(TokenUtility.CreateLiteralToken(assignment.Expression));
        }
    }

    private static void WriteResolvedExpression(SyntaxNode target, Expression expression)
    {
        if (expression is UnaryExpression {NodeType: ExpressionType.Convert} unary)
        {
            WriteResolvedExpression(target, unary.Operand);
            return;
        }

        if (expression is BinaryExpression binary)
        {
            WriteResolvedExpression(target, binary.Left);
            target.AddChild(Operators.Or);
            WriteResolvedExpression(target, binary.Right);
            return;
        }

        if (expression.Type.IsEnum)
        {
            TypeUtility.WriteName(target, expression.Type);
            target.AddChild(Operators.Dot);
        }

        target.AddChild(TokenUtility.CreateLiteralToken(expression));
    }
}

File: Languages/SyntaxGenDotNet.VisualBasic/Utilities/AttributeUtility.vb
﻿Imports System.Collections.ObjectModel
Imports System.Linq.Expressions
Imports System.Reflection
Imports SyntaxGenDotNet.Attributes
Imports SyntaxGenDotNet.Syntax
Imports SyntaxGenDotNet.Syntax.Tokens

Namespace Utilities
    ''' <summary>
    '''     Provides utility methods for working with attributes in the Visual Basic language.
    ''' </summary>
    Public Module AttributeUtility
        ''' <summary>
        '''     Writes a custom attribute to the specified node.
        ''' </summary>
        ''' <param name="target">The node to which to write the attribute.</param>
        ''' <param name="attributeExpression">The attribute expression to write.</param>
        Public Sub WriteCustomAttribute(target As SyntaxNode, attributeExpression As MemberInitExpression)
            Dim options As New TypeWriteOptions With {
                    .WriteAlias = False,
                    .WriteNamespace = True,
                    .TrimAttributeSuffix = True
                    }

            ' explicit creation of the open bracket token to avoid the trailing whitespace
            ' of the previous token being trimmed, as Operators.OpenBracket defaults to trimming.
            Dim openBracket As New OperatorToken(OpenChevron.Text, False)

            target.AddChild(openBracket)
            WriteAlias(target, attributeExpression.Type, options)

            Dim arguments As ReadOnlyCollection(Of Expression) = attributeExpression.NewExpression.Arguments
            Dim hasArguments As Boolean = arguments.Count > 0
            Dim hasBindings As Boolean = attributeExpression.Bindings.Count > 0

            If hasArguments Or hasBindings Then
                target.AddChild(OpenParenthesis)

                If hasArguments Then
                    WriteArguments(target, arguments)
                End If

                If hasBindings Then
                    WriteBindings(hasArguments, target, attributeExpression, options)
                End If

                target.AddChild(CloseParenthesis)
            End If

            target.AddChild(CloseChevron.With(Sub(o) o.TrailingWhitespace = WhitespaceTrivia.NewLine))
        End Sub

        ''' <summary>
        '''     Writes all known supported custom attribute to the specified node.
        ''' </summary>
        ''' <param name="generator">The syntax generator.</param>
        ''' <param name="target">The node to which to write the attributes.</param>
        ''' <param name="member">The member whose custom attributes to write.</param>
        Public Sub WriteCustomAttributes(generator As SyntaxGenerator, target As SyntaxNode, member As MemberInfo)
            For Each writer As AttributeExpressionWriter In generator.AttributeExpressionWriters
                Dim attributeType As Type = writer.AttributeType
                Dim attributes As Attribute() = member.GetCustomAttributes(attributeType, False).OfType (Of Attribute)().ToArray()
                Dim expressions As IEnumerable(Of Expression) = writer.CreateAttributeExpressions(member, attributes)

                For Each expression As Expression In expressions
                    If expression.NodeType <> ExpressionType.MemberInit Then
                        Continue For
                    End If

                    Dim memberInitExpression = DirectCast(expression, MemberInitExpression)
                    WriteCustomAttribute(target, memberInitExpression)
                Next
            Next
        End Sub

        Private Sub WriteArguments(target As SyntaxNode, arguments As ReadOnlyCollection(Of Expression))
            For index = 0 To arguments.Count - 1
                Dim argument As Expression = arguments(index)
                WriteResolvedExpression(target, argument)
            Next
        End Sub

        Private Sub WriteBindings(writeComma As Boolean,
                                  target As SyntaxNode,
                                  expression As MemberInitExpression,
                                  options as TypeWriteOptions)
            Dim comma As SyntaxNode = Operators.Comma.With(Sub(o) o.TrailingWhitespace = WhitespaceTrivia.Space)
            Dim bindings As ReadOnlyCollection(Of MemberBinding) = expression.Bindings
            options.WriteNamespace = False

            For index = 0 To bindings.Count - 1
                Dim binding As MemberBinding = bindings(index)
                If binding.BindingType <> MemberBindingType.Assignment Then
                    Continue For
                End If

                Dim assignment = DirectCast(binding, MemberAssignment)

                If index > 0 Or writeComma Then
                    target.AddChild(comma)
                End If

                target.AddChild(New IdentifierToken(binding.Member.Name))
                target.AddChild(Colon)
                target.AddChild(Operators.Assignment)

                If assignment.Expression.Type.IsEnum Then
                    WriteName(target, assignment.Expression.Type, options)
                    target.AddChild(Dot)
                End If

                target.AddChild(CreateLiteralToken(assignment.Expression))
            Next
        End Sub

        Private Sub WriteResolvedExpression(target As SyntaxNode, expression As Expression)
            Dim unary = TryCast(expression, UnaryExpression)
            If unary IsNot Nothing
                If unary.NodeType = ExpressionType.Convert Then
                    WriteResolvedExpression(target, unary.Operand)
                    Return
                End If
            End If

            Dim binary = TryCast(expression, BinaryExpression)
            If binary IsNot Nothing Then
                WriteResolvedExpression(target, binary.Left)
                target.AddChild(OrKeyword)
                WriteResolvedExpression(target, binary.Right)
                Return
            End If

            If expression.Type.IsEnum Then
                WriteName(target, expression.Type)
                target.AddChild(Dot)
            End If

            target.AddChild(CreateLiteralToken(expression))
        End Sub
    End Module
End Namespace
